### ISSUE TITLE
question on backward in shards

### ISSUE THREAD

**Issue by moonlightlane** [OpenNMT/OpenNMT-py#387]
I notice there are two backward called, one before the loss calculation and one after. The one before loss calculation is in `shards` line 185 in `Loss.py`, and the backward after loss calculation is in `sharded_compute_loss` in line 74 in `Loss.py`. I thought I only need to do the backward step one time, on the loss, after it is calculated? What is the previous backward step doing?

I understand using shard is for memory efficiency. But suppose I do not use shards. Then in each batch, after forward pass of the encoder, decoder and generator, I simply do the following:

`criterion = nn.NLLLoss() # defined elsewhere, in train.py perhaps`
`...`
`loss = criterion(out, target)`
`loss.div(batch.batch_size).backward()`

where `out` is the output from the generator.

This should work just as well, correct?

**Comment by ketranm** [comment]
your code should work.
For the first question, it's a bit complicated. The first backward pass does not backprop through the whole network, but instead it is used to compute the gradient of the loss w.r.t the output of the network.  See the `filter_shard_state(state)` function, it basically creates a placeholder for this gradient.
After the last shard is processed, we collect all the gradient of the loss w.r.t the output and then backprop through entire network.

**Comment by moonlightlane** [comment]
@ketranm Thanks, I think I understand better now! Let me try to restate what you said to make sure I understand: the shards divide the loss computation in each batch into yet smaller pieces, by processing a smaller number of words in each batch. For example, the default shard size is 32. Assume the maximum target length is 100, then loss will be computed in 4 shards. In each shard, the loss of each word in that shard is computed, and the gradient is accumulated wrt the last layer (which is a softmax). When all shards finish computing loss, it means that the loss in the current batch has been computed, and then we do a backward pass through the entire network to get all the gradients.

So I guess I can indeed just compute and accumulate loss with all the words in the batch, and the do a one-time backward pass on this accumulated loss to compute all the gradients.

**Comment by xubenben** [comment]
@moonlightlane loss.backwards execute first, line 185 execute when all shards loss have been processed.

**Comment by xubenben** [comment]
@moonlightlane there is a small trick in the code.  

- variable "output" computed in loss.backowrd() is filtered in function filter_shard_state, which means "output" = Varaible("output",...)

-  variable "output" computed in line 185 hasn't been filtered which means backward will bp to whole computation graph while loss.backowrd() will not

**Comment by ketranm** [comment]
@moonlightlane Set `max_generator_batches` to a very large value would do the job. However, i think using `nn.NLLLoss` is better in this case
```
weight = torch.ones(vocab_size)
weight[padding_idx] = 0
crit = nn.NLLLoss(weight, size_average=False)
```

**Comment by moonlightlane** [comment]
@xubenben @ketranm Thanks for the nice clarification!! I also verified this using `register_hook` at context vector output by the encoder. The hook only prints on line 185 which means that line 185 is where the entire graph is backprop'd.

**Comment by KelleyYin** [comment]
I can't completely understand its means of the backward in line 185 . I do some changes when calculating the loss. changing the  loss function . However, it will arise a error: there are no graph nodes that require computing gradients. I convince the backward in line 185 lead to generate the bug . Therefore ,I want to completely understand the function.

**Comment by memeda** [comment]
@KelleyYin 
I think I figure it out.
the code has changed and line number can't match,  I paste the  2 backwards  code follows:

1. in LossComputeBase, after get every share,  do a backward, that is:

```Python
        for shard in shards(shard_state, shard_size):
            loss, stats = self._compute_loss(batch, **shard)
            loss.div(float(normalization)).backward()
            batch_stats.update(stats)
```

2. in `shards(state, shard_size, eval_only=False)`, after yiled the last share, do a backward from shared to input, that is

```Python
        for k, (v, v_split) in non_none.items():
            if isinstance(v, torch.Tensor) and state[k].requires_grad:
                variables.extend(zip(torch.split(state[k], shard_size),
                                     [v_chunk.grad for v_chunk in v_split]))
        inputs, grads = zip(*variables)
        torch.autograd.backward(inputs, grads)
```
----

Now I describe it:

1.  split the shares,  and **unlink** the share from original network
       see code `filter_shard_state(state, shard_size=None)`, it use `v_chunk.data.clone()` to clone the original chunk data, instead of `directly assign`, so it unlink the data from the original graph.
       That is, the following `backward` on this data **can't backwards to the original graph!!**

2.  calculate the gradient,  the loss w.r.t the `unlinked shared data`
     in code `LossComputeBase`'s backwards, it does. As previous says, The loss gradients can only pass to the share data, not the original graph.

3. use the grad from the `unlink shared data`  to backward to the original graph
    in code `shards(state, shard_size, eval_only=False)`, it does.
    see the code 
    ```
                variables.extend(zip(torch.split(state[k], shard_size),
                                     [v_chunk.grad for v_chunk in v_split]))
    ```
    
     Clearly, it `re-split` the data from the original graph(to get the link to the original graph) , and bind the grads corresponding to the `unlink shared data`. 
     so after binded, here the `backard(inputs, grads)` do the really backward for the original Graph.

----

So Why do this? I think is for efficiency.

the trival netork can viewed as:  ENCODER -> OUTPUT -> LOSS, and grads: LOSS -> OUTPUT -> ENCODER. 

but here we split the `OUTPUT` to multiple shares. that is

```
ENCODER  |-> OUTPUT_SHARE1 -> LOSS1
                    |-> OUTPUT_SHARE2 -> LOSS2
                    |-> OUTPUT_SHARE3 -> LOSS3
```

if we do backward for every share,  we will got backwards caluculation:

```
LOSS1 -> OUTPUT_SHARE1 -> ENCODER
LOSS2 -> OUTPUT_SHARE2 -> ENCODER
LOSS3 -> OUPTUT_SHARE3 -> ENCODER
```

so the `OUTPUT* -> ENCODER` is duplicated and time-consuming.
